### ISSUE TITLE
BAT_CONFIG_PATH ignored by --generate-config-file

### ISSUE THREAD

**Issue by DH-Illidan** [sharkdp/bat#1550]
**What version of `bat` are you using?**

Newest

**Describe the bug you encountered:**
![image](https://user-images.githubusercontent.com/55729509/108675764-3694ee80-7522-11eb-9c45-3337c8d63dc7.png)

**What did you expect to happen instead?**

use custom location instead

**Windows version:**

![image](https://user-images.githubusercontent.com/55729509/108675495-d736de80-7521-11eb-81ef-032ba4309f1f.png)

**Comment by sharkdp** [comment]
Thank you for reporting this. Looks like a big indeed.

**Comment by Enselic** [comment]
I can confirm this bug, which is generic. It is caused by
```
        .filter(|config_path| config_path.is_file())
```
in
```
pub fn config_file() -> PathBuf {
    env::var("BAT_CONFIG_PATH")
        .ok()
        .map(PathBuf::from)
        .filter(|config_path| config_path.is_file())
        .unwrap_or_else(|| PROJECT_DIRS.config_dir().join("config"))
}
```
which makes sense when _reading_ config values, but of course not when _creating it_.

Made a proof of concept fix (see above), but am not happy with how it turned out. Plus, it needs integration tests. Will take care of that at some later point when I get time, unless someone beats me to it (which I encourage).

**Comment by sharkdp** [comment]
> which makes sense when _reading_ config values, but of course not when _creating it_.

Let's discuss this first. Maybe it doesn't make sense for _reading_ either.

Agreed: when *writing* with `--generate-config-file`, we should always simple write to the path returned by `config_file()` (maybe rename that function to `config_file_path()`). We should print an error if that path can not be written to (e.g. missing permission or path exists and is a directory).

When *reading* the config file, we could distinguish between two cases:
1. `BAT_CONFIG_PATH` has not been set => read from the default path, but ignore if it is not existent. Print an error in case of permission / file-type problems.
2. `BAT_CONFIG_PATH` has been set => I think we should always try to read from that path and maybe print an error/warning if it does not exist. Maybe we should even ignore it, just like in case 1. In case of permission / file-type problems, we print an error.

I guess `config_file` would be most useful if it always just returns the path where `bat` expects the config to be. Even if that path does not exist. That would mean that we could simply remove the `.filter(…)` call.

**Comment by sharkdp** [comment]
Bugfix released in [bat v0.18](https://github.com/sharkdp/bat/releases/tag/v0.18.0).
